src/blas/impl/KokkosBlas1_iamax_impl.hpp

```diff
@@ -0,0 +1,231 @@
+        mag_type val = IPT::norm (m_x(i,lid));
+        if( val > thread_lmaxloc.val ) { thread_lmaxloc.val = val; thread_lmaxloc.loc = i; }      
+      }, maxloc_reducer(col_maxloc));
+      if (tid==0) m_r(lid) = col_maxloc.loc;
+    }
+};
+
+
+/// \brief Find the index of the element with the maximum magnitude of the single vector (1-D
+///   View) X, and store the result in the 0-D View r.
+template<class RV, class XV, class SizeType>
+void
+V_Iamax_Invoke (const RV& r, const XV& X)
+{
+  typedef typename XV::execution_space execution_space;
+  typedef Kokkos::Details::ArithTraits<typename XV::non_const_value_type> AT;
+  typedef typename AT::mag_type mag_type;
+
+  typedef typename Kokkos::MaxLoc<mag_type,SizeType>::value_type maxloc_type;
+  typedef Kokkos::MaxLoc<mag_type,SizeType> maxloc_reducer;
+
+  const SizeType numRows = static_cast<SizeType> (X.extent(0));
+
+  // Avoid MaxLoc Reduction if this is a zero length view
+  if( numRows == 0 ) {
+    Kokkos::deep_copy(r,0);
+    return;
+  }
+
+  Kokkos::RangePolicy<execution_space, SizeType> policy (0, numRows);
+
+  typedef V_Iamax_Functor<RV, XV, mag_type, SizeType> functor_type;
+  functor_type op (X);
+  maxloc_type maxloc;
+  Kokkos::parallel_reduce ("KokkosBlas::Iamax::S0", policy, op, maxloc_reducer(maxloc));
+  r() = maxloc.loc;  
+}
+
+
+/// \brief Find the index of the element with the maximum magnitude of the columns of the
+///   multivector (2-D View) X, and store result(s) in the 1-D View r.
+template<class RV, class XMV, class SizeType>
+void
+MV_Iamax_Invoke (const RV& r, const XMV& X)
+{
+  typedef typename XMV::execution_space execution_space;
+  typedef Kokkos::Details::ArithTraits<typename XMV::non_const_value_type> AT;
+  typedef typename AT::mag_type mag_type;
+
+  const SizeType numRows = static_cast<SizeType> (X.extent(0));
+
+  // Avoid MaxLoc Reduction if this is a zero length view
+  if( numRows == 0 ) {
+    Kokkos::deep_copy(r,0);
+    return;
+  }
+
+  typedef Kokkos::TeamPolicy<execution_space>  team_policy;
+  const team_policy policy( r.extent(0), Kokkos::AUTO );
+
+  // If the input multivector (2-D View) has only one column, invoke
+  // the single-vector version of the kernel.
+  if (X.extent(1) == 1) {
+    typedef Kokkos::View<typename RV::non_const_value_type,
+                         typename RV::array_layout,
+                         typename RV::device_type,
+                         Kokkos::MemoryTraits<Kokkos::Unmanaged>> RV0D;
+    RV0D r_0(r, 0);
+    auto X_0 = Kokkos::subview (X, Kokkos::ALL (), 0);
+    typedef decltype (X_0) XV1D;
+    V_Iamax_Invoke<RV0D, XV1D, SizeType> (r_0, X_0);
+  }
+  else {
+    typedef Kokkos::View<typename RV::non_const_value_type *,
+                         typename RV::array_layout,
+                         typename XMV::device_type> RV_D;
+    typedef MV_Iamax_FunctorVector<RV_D, XMV, mag_type, SizeType> functor_type;
+    RV_D r_d("r_d", r.extent(0));
```

Comment:
Why do we need to allocate temporary storage here?  Is the issue that `r` must be host storage?

BLAS `IxAMAX` functions only work with 1-D input arrays, so unless we have a use case for the 2-D case, why not just get rid of it and thereby reduce code complexity?

Reply:
@mhoemmen yes,  ```r``` is in host space. But, actually, we do not have a use case for the 2-D. I just added this for the sake of completeness. Of course, I can remove this 2-D case.

Reply:
The 2-D case could be useful for panel factorizations, but in that case, `r` would always be on device.

Reply:
@mhoemmen So, I need to change the multi-vector template such that ```r``` is on device. I wasn't sure if ```r``` should be on device or on host.

Reply:
As long as these functions never allocate Views, I'm happy.  You have a View, so you can check its memory space accessibility.  The 1-D input case can either return a value (on host) or write to a 0-D View (on device).

Reply:
Thanks @mhoemmen. The code is fixed accordingly. Basically, there is no view allocation.